### ISSUE TITLE
Deployed Angular 5 app results in 404 for require.js

### ISSUE THREAD

**Issue by chalin** [dart-lang/build#896]
TL;DR `<base href>` is ignored by `main.dart.js` generated for Angular 5 app.

---

The deployed version of the 5.0.0-alpha+4 version of the [quickstart][] example app fails to load with the following **error appearing in the console**:

```nocode
GET http://angular-examples.github.io/packages/$sdk/dev_compiler/web/dart_stack_trace_mapper.js net::ERR_ABORTED
(anonymous) @ main.dart.js:50

GET http://angular-examples.github.io/packages/$sdk/dev_compiler/amd/require.js net::ERR_ABORTED
(anonymous) @ main.dart.js:56
```

To reproduce, visit: http://angular-examples.github.io/quickstart/5

This problem did not appear using the old build system under Angular 4.

What seems to be happening is that the `<base href="/quickstart/5/">` contained in `index.html` is being ignored by by [main.dart.js](https://github.com/angular-examples/quickstart/blob/gh-pages/5/main.dart.js#L32-L54):

```js
  var baseUrl = ...
  var el;
  el = document.createElement("script");
  ...
  el.src =
    baseUrl + "packages/$sdk/dev_compiler/web/dart_stack_trace_mapper.js";
 ...
```

The setting of the baseUrl doesn't seem to take the `<base href>`.

cc @kwalrath @kevmoo @matanlurey 

[quickstart]: https://webdev-dartlang-org-dev.firebaseapp.com/angular/guide/setup#create-a-starter-project

**Comment by matanlurey** [comment]
The example you are showing is with `dartdevc`, did you used to do that with v4?

**Comment by chalin** [comment]
No, under v4 we used dart2js.

**Comment by matanlurey** [comment]
So that's not comparing apples and oranges then :)

Any reason you'd want to _deploy_ `dartdevc`? That usually isn't desired.

**Comment by chalin** [comment]
Sure, the goal wasn't to compare but to point out the issue.

> Any reason you'd want to deploy dartdevc? That usually isn't desired.

Jan 17, 18:13, on Gitter [dart-lang/source_gen](https://gitter.im/dart-lang/source_gen), we were discussing the building and testing of the Angular example apps. In the end you wrote:

> There probably isn’t huge value in running in dart2js
> We don’t internally

To which Jacob agreed. Until there is an easier way to switch between web compilers (#801), I was planning on building and deploying using dartdevc.

In any case, if this is a real problem with dartdevc, should I be reporting the issue over dartdevc?

**Comment by matanlurey** [comment]
@chalin:

> _deploying using dartdevc_

This definitely isn't desirable, for lots of reasons. We don't do this internally.

> In any case, if this is a real problem with dartdevc, should I be reporting the issue over dartdevc?

It's not clear to me if this is an issue with `dartdevc` or not. I'll let @jakemac53 weigh in.

**Comment by chalin** [comment]
My bad, I misunderstood. I'll switch back to deploying using dart2js.

**Comment by jakemac53** [comment]
The issue here is that it is being deployed under a subdirectory, we don't really have a good solution for that at this time with dartdevc. We should keep this issue open anyways as actually solving it would also allow us to remove some hacks that are currently in place for package:test.

**Comment by chalin** [comment]
Thanks for the feedback @jakemac53.

@matanlurey asked:

> The example you are showing is with dartdevc, did you use to do that with v4, or dart2js?

I've tested this locally and I can confirm that this problem doesn't manifest itself when using `dart2js`.

Note that the script paths created by the default dart2js transformer (or whatever a "transformer" is called under the new build system), have no leading `/`, e.g.:

```html
<script defer src="packages/browser/dart.js"></script>
```

As you probably know, the dynamically generated path to, say, `require.js` _does_ start with a `/`, e.g.:

```html
<script defer="" src="/packages/$sdk/dev_compiler/amd/require.js" data-main="http://angular-examples.github.io/quickstart/5/main.dart.bootstrap"></script>
```

(1) So one thing would be to have the dynamic `baseUrl` computation use "" as a default path rather than `/`.

(2) The other problem seems to be that required scripts like the following are being dynamically added to the top of the `<head>` element, **_before_** the `<base href>`:

```html
<script type="text/javascript" charset="utf-8" async="" data-requirecontext="_" data-requiremodule="main.dart.bootstrap" src="http://localhost:8080/web/main.dart.bootstrap.js"></script>
```

These are `async`, so couldn't they be added at the end of head? If not, I suppose it is ok for them to be added to the top, but if they are, then any `<base href>` element should then be relocated to the top too.

WDYT?

**Comment by chalin** [comment]
FYI, the app at http://angular-examples.github.io/quickstart/5 has been replaced by the dart2js version so you won't be able to see a hosted version of the dartdevc-built app anymore.

**Comment by matanlurey** [comment]
Any actionable item here @chalin?

**Comment by matanlurey** [comment]
Great!

**Comment by davidmorgan** [comment]
Closing as stale.
